server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java

```diff
@@ -1011,19 +1030,33 @@ private void setUpdateTablet(UpdateSession us, KeyExtent keyExtent) {
         updateMetrics.addPermissionErrors(0);
         return;
       }
+
     }
 
     @Override
     public void applyUpdates(TInfo tinfo, long updateID, TKeyExtent tkeyExtent,
-        List<TMutation> tmutations) {
+        List<TMutation> tmutations) throws TException {
       UpdateSession us = (UpdateSession) sessionManager.reserveSession(updateID);
       if (us == null) {
         return;
       }
 
+      Semaphore semaphoreCopy = getSemaphore();
       boolean reserved = true;
+      boolean semaphoreAcquired = false;
       try {
         KeyExtent keyExtent = new KeyExtent(tkeyExtent);
+
+        if (TabletType.type(keyExtent) == TabletType.USER) {
+          if (!semaphoreCopy.tryAcquire()) {
+            throw new TException("Mutation failed. No threads available.");
```

Comment:
We could consider having a specialized thrift exception that is thrown to indicate resource exhaustion on tserver.  When the batch writer sees this it could retry with exponential backoff up to its configured time-out.